[mlir] Add optional layout attribute to VectorType

harsh-nod opened the pull request:
This patch adds an attribute interface for representing the layout on vector types. This layout could be used to represent the mapping from the vector indices to the indices of the vector fragments held by different threads of a GPU.

The interface has a verify function that can be used to validate that the layout accurately represents the vector shape.

harsh-nod committed:
[mlir] Add optional layout attribute to VectorType

This patch adds an attribute interface for representing the
layout on vector types. This layout could be used to represent
the mapping from the vector indices to the indices of the
vector fragments held by different threads of a GPU.

The interface has a verify function that can be used to validate
that the layout accurately represents the vector shape.

rengolin reviewed (commented):
The code looks ok to me, the abstraction level to keep it implementation defined, too. Should be fine for the hands-off approach we're having now.

I wonder if we want to have some implementation upstream, though. Sure, this would be for another PR, but how many upstream passes are we going to disable because they don't understand the downstream layout encoding?

I'll leave for others to finally approve the PR, to give them time to review.

ThomasRaoux commented on mlir/lib/AsmParser/TypeParser.cpp:
```diff
@@ -459,14 +459,38 @@ VectorType Parser::parseVectorType() {
+    return success();
+  };
+
+  // Parse a list of mappings and address space if present.
+  if (!consumeIf(Token::greater)) {
+    // Parse comma separated list of affine maps, followed by memory space.
```
those comments look out of date.

harsh-nod replied:
Thanks, will update.

ThomasRaoux commented on mlir/lib/IR/BuiltinTypes.cpp:
```diff
@@ -242,6 +243,11 @@ LogicalResult VectorType::verify(function_ref<InFlightDiagnostic()> emitError,
 
+  if (layout) {
+    if (failed(layout.verifyLayout(shape, elementType, emitError)))
+      return emitError() << "layout does not match underlying vector shape";
```
the error message should be more generic as the mismatch may be in the shape or somewhere else.

harsh-nod replied:
Makes sense, will change to something more generic like "Layout verification failed!"

ThomasRaoux commented on mlir/unittests/Interfaces/VectorLayoutInterfaceTest.cpp:
```diff
@@ -0,0 +1,158 @@
+};
+
+TEST(VectorLayoutAttrInterface, NamedStridedLayout) {
+  const char *ir = R"MLIR(
```
can you also add a lit test to test round trip printing/parsing?

harsh-nod replied:
sure will add a roundtrip test. Might have to keep it as a unit test since the attribute and dialect are only defined in this cpp file.

ThomasRaoux commented on mlir/include/mlir/IR/BuiltinTypes.td:
```diff
@@ -1050,6 +1052,14 @@ def Builtin_Vector : Builtin_Type<"Vector", [ShapedTypeInterface], "Type"> {
 
+    ##### Layout
+
+    A vector may optionally have a layout that indicates how indices of
+    the vector are transformed to indices of the vector fragments that
+    are held by individual threads in a SIMT execution model. Such layouts
+    are common in a wide variety of GPU matrix multiplication instructions.
+    The layout can be any attribute that implements `VectorLayoutAttrInterface`.
```
I think the description should probably be more generic than that to allow for more usage of the layout. (not only GPU and mapping to SIMT registers)

harsh-nod replied:
Sure, will rewrite this to make it more general. I think the primary motivation for this was GPU/SIMT but people can use this in different ways.

ThomasRaoux reviewed (commented):
Does that mean that every vector transformation is now responsible for maintaining and propagating this attribute? That seems like a fairly fundamental change, do we need to make sure the community is onboard with that? What's going to be the expected behavior of transformations regarding this new attribute in patterns handling vectors?

dcaballe commented:
Hey Hash, thanks for the proposal. This is an interesting change! Some initial questions that come to mind:
* How would we lower something like this to LLVM?
* Should this be a property of the type or of the operations supporting these swizzles? In the doc you mention that this is mostly for SIMT matmuls, which make me wonder if we should only make this available for those ops.
* Assuming this should be a property of the type, should we consider a "LayoutVectorType", where the attribute is not optional but mandatory?
* It would be very helpful to add a bunch for IR examples. I would help me get a clearer idea of what we are trying to achieve here.

This looks like a big change so I would give it enough time to settle. Perhaps we should discuss some of these questions around an RFC?

harsh-nod commented:
> Does that mean that every vector transformation is now responsible for maintaining and propagating this attribute? That seems like a fairly fundamental change, do we need to make sure the community is onboard with that? What's going to be the expected behavior of transformations regarding this new attribute in patterns handling vectors?

I don't think every vector transformation should be responsible for maintaining and propagating this attribute. The existing vector transformations and patterns can safely ignore this attribute which hopefully reduces the scope of this change.

harsh-nod commented:
> Hey Hash, thanks for the proposal. This is an interesting change! Some initial questions that come to mind:
> 
> * How would we lower something like this to LLVM?
> * Should this be a property of the type or of the operations supporting these swizzles? In the doc you mention that this is mostly for SIMT matmuls, which make me wonder if we should only make this available for those ops.
> * Assuming this should be a property of the type, should we consider a "LayoutVectorType", where the attribute is not optional but mandatory?
> * It would be very helpful to add a bunch for IR examples. I would help me get a clearer idea of what we are trying to achieve here.
> 
> This looks like a big change so I would give it enough time to settle. Perhaps we should discuss some of these questions around an RFC?

So this PR is un-opinionated in terms of how the vector layout can be represented. The idea behind that is that many people will have a preference on what layout representation best suits them and we would like upstream to be able to support everyone's ideas. So regarding lowering to LLVM, this PR does not prescribe a way to utilize the layout information to lower to LLVM. That will be user dependent. That being said, I am a proponent of a specific form of the layout representation that can be used to lower to LLVM. For more details on how that works as well as IR examples, please see the following slides and videos -
https://github.com/nod-ai/techtalks/blob/main/high_dimensional_layout_flash_attn_harsh_23.pdf
https://www.youtube.com/watch?v=fqeyTCmqO4g
https://www.youtube.com/watch?v=5i7xrBUCD38

Regarding whether this should be a property of the type or the operations that support these swizzles -
I believe this should be a property of the type but it stems from constraints/requirements on the operations. While my work is currently mostly driven by GPU/SIMT requirements, I believe this layout is more general than that and could be used for other execution models as well.

Regarding a new vector with layout type, that is an interesting idea. However, my concerns would be the scope of the changes required to have all the vector, arithmetic etc. ops support this new type. Also, this new type would be a superset of the existing vector type so would seem like a lot of duplication.

Hope the links above provide you with enough IR examples. If not, happy to get on a GVC to give more info. 

Happy to continue the conversation anywhere. I am hoping that this is not a big change with the goal that everyone should hopefully be able to continue using vector ops and types as usual with extra work required from those who want to use this layout.

MaheshRavishankar commented:
The way I see it, this is part of the type (like memory space is part of memref type). For most transformation this should be a pass through. Some transformations (in-tree or out-of three) can implement transformations that use these attributes. As Harsh mentioned I dont expect anything to directly work on this Attribute type, but this is the parent class for all derived attribute types for specific use cases..
There is going to be the issue of making sure transformations dont drop the attribute from the type, but that is probably something worth doing anyway.

harsh-nod committed:
Address Thomas' comments

- Fix out of date comments
- More generic error message on layout verification failure
- Add roundtrip test
- More generic description of layout

lattner reviewed (changes requested):
This patch extends the Vector type but does not update any of the existing clients to handle it (defensively or otherwise).  It'd be great to come to consensus of whether extending this at all is a good idea:
https://discourse.llvm.org/t/rfc-remove-arith-math-ops-on-tensors/74357/44

if it is a good idea, then all the existing clients should be updated.

nicolasvasilache commented:
This needs its own RFC, I too would prefer to see something less intrusive with room for evolution.

One possible path forward is to start from something like a FragmentedVectorType that is not automatically a builtin VectorType and interfaces that define the semantics of such fragments. The existing VectorType is a particular limit case with a single fragment.

Over time, ops, patterns and transformations can be ported to `VectorTypeOrFragmentedVectorType`.
At some point in the future, if we get to full convergence great, if not that's fine too: it is unclear that everything makes sense to have as a fragmented vector.

One big question is whether the number of fragments and the data paths to convert between different such "layouts" are statically known but the system should be designed such that we can ask this question later (e.g. if `?` quantities in vectors ever become a thing).
